### ISSUE TITLE
APFS and MacOS 10.13 High Sierra

### ISSUE THREAD

**Issue by gondalez** [docker/for-mac#1810]
I am just curious if anyone (contributors or otherwise) has tried docker for mac with the dev beta of High Sierra that ships with the new APFS.

A few burning questions:
- does it work?
- is it performant?
- can APFS help with #1592?

**Comment by ChefAustin** [comment]
Last night I installed the most recent stable release of Docker for Mac (17.06.0-ce Build 02c1d87) on an APFS formatted disk with macOS 10.13 Beta (17A291m). Docker installed/started properly, and would even pull down container images from Dockerhub, but upon trying to `docker start <container_id>` the container would start, then immediately stop, and attempt a restart (to no avail).
Same behavior/issue elicited in #1838

**Comment by burnedikt** [comment]
Just tested Docker (17.06.0-ce-mac19 (18663)) with APFS on my new Macbook Pro running Mac OS High Sierra (10.13 Beta (17A306f) and it seems to work for me as expected. Installation and startup of Docker for Mac itself without any problems. Also started some existing containers successfully and pulled a couple of existing images to fire them up, e.g. `docker run -it --rm busybox`or `docker run -it --rm node`. So far, everything seems to work as expected for me.

**Comment by maurosbu** [comment]
@burnedikt What about performances? Is https://github.com/IFSight/d4m-nfs still required?
Many thanks!

**Comment by burnedikt** [comment]
@maurosbu I'll gladly try it out, if you can give me a sample container for benchmarking

**Comment by maurosbu** [comment]
@burnedikt Is file access in mounted volumes still extremely slow (https://github.com/docker/for-mac/issues/77)?

**Comment by pfx2** [comment]
I’m on High Sierra GM with APFS and volumes are still slow as hell...
But maybe the Docker team will be able to improve the performance of volumes based on APFS features. We have to give them time as APFS is not documented well.

**Comment by EugenMayer** [comment]
Was really hoping for this, seems i have to continue with http://docker-sync.io for even longer time ;/

**Comment by kingbuzzman** [comment]
Whats the status on this?

**Comment by qrpike** [comment]
It's PAINFULLY slow if you're mounting directories to your host machine... and I do mean painfully slow. Takes several minutes to start a redis or mongo where the files are mounted on host machine ( very small DB ).

**Comment by tcalin** [comment]
Define PAINFULLY!

**Comment by barat** [comment]
Someone tested newest edge (raw instead of qcow2 on apfs High Sierra)? I cannot test yet - need to finish&push a task before experiments :)

**Comment by gondalez** [comment]
Thanks for the tip @barat I hadn't seen that release. Sounds very promising! I'm mid-task also but will try is as soon as I'm done :)

To save some time for others, from the [17.11.0-ce-rc4-mac39 release notes](https://docs.docker.com/docker-for-mac/release-notes/#edge-release-notes):

>For systems running APFS on SSD on High Sierra, use raw format VM disks by default. This increases disk throughput (from 320MiB/sec to 600MiB/sec in dd on a 2015 MacBook Pro) and disk space handling.
...

**Comment by maparent** [comment]
I can confirm that it feels much faster than before, though I do not have figures.

**Comment by richardfickling** [comment]
Upgrading from main line to edge (17.11.0-ce-mac40 (20561)) brought our test run time from 150 seconds down to 22. Most of the time spent in those tests is writing/reading from a mysql container.

**Comment by mauritsl** [comment]
Can also confirm that things went a lot faster on 17.11.0-ce-mac40. It's also visible in a simple dd test:

```
docker run -it --rm ubuntu:16.04 dd if=/dev/zero of=/root/testfile bs=512 count=1000 oflag=direct
```

On stable @ high sierra it shows 2.7 MB/s, with edge 4.8 MB/s.

**Comment by EugenMayer** [comment]
@mauritsl since i do not have / plan to have high sierra, could you help me with your benchmark?

i would love to fill the tabel here https://github.com/EugenMayer/docker-sync/wiki/4.-Performance with the result here - to be transparent if docker-sync has any value or how it compares.

Could you run https://github.com/EugenMayer/docker-sync/wiki/Performance-Tests-2017#docker-for-mac---native .. ensure to fix the mount path. And also ensure to notice, that that is missing on your test you are not using any mount, so no OSXFS is involved, that should be native speed anyway. 

That would already be great and should you not take more then 60 seconds.

If you feel luck, i would love to see how the docker-sync tests performance OSXFS based ( native_osx strategy ) https://github.com/EugenMayer/docker-sync/wiki/Performance-Tests-2017#docker-sync---strategy-native_osx

That should not take you more then 2 minutes
```
gem install docker-sync
git clone https://github.com/EugenMayer/docker-sync-boilerplate
cd docker-sync-boilerplate/default
docker-sync-stack start

docker exec -it nativeosx_app-unison_1 time dd if=/dev/zero of=/var/www/test.dat bs=1024 count=100000
``` 

I am sorry i cannot do this myself - but i currently cannot upgrade on High Sierra

**Comment by mauritsl** [comment]
@EugenMayer here is the output for these benchmarks. Running on MacBook Pro (Retina, 13-inch, Late 2013) with 2.8 GHz Intel Core i7, 16GB RAM and SSD. (3 cores / 6GB RAM assigned to Docker).
```
➜  ~ docker run -it -v /Users/maurits/vhosts:/var/www alpine time dd if=/dev/zero of=/var/www/test.dat bs=1024 count=100000
100000+0 records in
100000+0 records out
real	0m 18.15s
user	0m 0.12s
sys	0m 0.76s

➜  default git:(master) ✗ docker exec -it default_app-native-osx_1 time dd if=/dev/zero of=/var/www/test.dat bs=1024 count=100000
100000+0 records in
100000+0 records out
real	0m 0.37s
user	0m 0.02s
sys	0m 0.29s

➜  unison git:(master) ✗ docker exec -it unison_app-unison_1 time dd if=/dev/zero of=/var/www/test.dat bs=1024 count=100000
100000+0 records in
100000+0 records out
real	0m 0.38s
user	0m 0.02s
sys	0m 0.33s
```
But Unison does not seem to be working. The stack does not show any changes if I edit the app/index.html. Other methods work fine.

**Comment by EugenMayer** [comment]
So @mauritsl thank you a lot - i added it on https://github.com/EugenMayer/docker-sync/wiki/4.-Performance.

For now, there is literally 0 improvement on the host mounted folder 18seconds vs 0.37seconds with docker-sync native speed.

So the improvement you see here is xhyve general I/O improvement on read/write - which is not bad not really the breaker, it was probably just 1.5x slower then linux - while  the host mounts are about 20-70x slower without docker-sync. 

Well, lets still hope :)

**Comment by vovimayhem** [comment]
Is anyone here on High Sierra, with APFS and Docker, using the `Docker.raw` file, and working with large-volume databases? I'm getting some random errors on a postgres container (for a data migration project I'm on), about corrupted blocks of data when using the `.raw` Disk image, but everything is OK when I revert to using `.qcow` disk image. (See #2327)

**Comment by devdude** [comment]
@vovimayhem I have been experimenting with Docker.raw, trying to get it to work with a 300 GB postgres db and have been having all kinds of random errors when trying to restore the 300GB db. FYI, I tried both the built in disk resize from Docker CE 17.12 as well as qemu-img with gparted. There seemed to be no issues when the postgres db was only 10 GB.

First I got a couple of these:
pg_restore: [archiver (db)] Error while PROCESSING TOC:
pg_restore: [archiver (db)] Error from TOC entry 11130; 2606 788502597 CONSTRAINT about_mes_pkey user_name
pg_restore: [archiver (db)] could not execute query: ERROR:  invalid page in block 7132 of relation base/21712/24758

Then:
pg_restore: [archiver (db)] Error from TOC entry 11132; 1259 1212121894 INDEX index_some_name_id username
pg_restore: [archiver (db)] could not execute query: server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.

And then hundreds of these errors:
pg_restore: [archiver (db)] Error from TOC entry 11134; 1259 1212121974 INDEX index_some_name_id username
pg_restore: [archiver (db)] could not execute query: no connection to the server

**Comment by vovimayhem** [comment]
@devdude On the bright side, it's reassuring to know I'm not alone. Those are the exact same issues I observe when using the `.raw` disk format.

**Comment by djs55** [comment]
@vovimayhem @devdude Do you have a repro case you could share (ideally one smaller than 300 GB)? It doesn't seem to happen in other environments, but perhaps postgres is able to hit some kind of corner case in the block device code (or in the APFS filesystem itself)

Does deleting images have anything to do with it? Docker for Mac only performs TRIM operations when images are deleted. I only ask because TRIM uses the new "punch hole in sparse file" feature in APFS.

**Comment by vovimayhem** [comment]
@djs55 In my case it may take me a while to come up with a database dump of my own with all the data scrambled to use as the input...

Anyways, when I got these problems, I'm certainly positive that I tried with the VM Disk blank (i.e. used the "Remove All Data" button in the "Reset" screen of Docker for Mac)

**Comment by djs55** [comment]
I wonder if it would be possible to create random data using a script which could trigger the problem? Do you think it matters what the schema is... do you think something simple would work?

**Comment by vovimayhem** [comment]
@djs55 what I noticed is that loading the data in the postgres server from a dump finishes with no apparent problems. But then, when I execute my migration script (which uses `INSERT SELECT` and `UPDATE + JOIN` queries to migrate the data into a new schema), is where the problems come up.

Now, reflecting on the "Docker for Mac only performs TRIM operations when images are deleted"... there is no image deletion between loading the dump and running the migration. What it **does** happen is a *container* removal - I use `docker-compose run --rm` commands to invoke each migration script. If the image and container removal are both using those `TRIM` operations, then maybe that's our culprit?

**Comment by daveisfera** [comment]
Postgres has restarted several times for me while just doing [`UPSERT`](https://wiki.postgresql.org/wiki/UPSERT) calls. No containers were starting or stopping at the time and the schema of the table was very simple.

**Comment by kingbuzzman** [comment]
Oh man.. this scares me enough into NOT updating :/ (I use postgres a lot...)

**Comment by EugenMayer** [comment]
There are a lot more issues with this and the new stable, FS events are los / stopping to propagate very fast https://github.com/EugenMayer/docker-sync/issues/517

**Comment by kingbuzzman** [comment]
@EugenMayer you're saying that a docker volume does not suffer from this issue? Only the volumes mapped back to the host?

**Comment by daveisfera** [comment]
In my case, the crashing of Postgres was happening when mounting a directory from the host.

**Comment by EugenMayer** [comment]
@kingbuzzman 99% of the problems mentioned here are only related to host mounts. And this is known for so long...

**Comment by vovimayhem** [comment]
@EugenMayer Mine is definitely **NOT** on a mounted folder. My use case is 100% docker volume stuff. I agree, anyone using postgres with data in a mounted host directory is asking for trouble.

**Comment by vovimayhem** [comment]
@kingbuzzman Everything works OK if using `*.qcow` instead of `*.raw` for the Docker VM disk... see https://github.com/docker/for-mac/issues/2327#issuecomment-351842252

**Comment by gsong** [comment]
I reset my local Docker environment today so I can test out the new `Docker.raw` format. I ran into couple issues:

1. I couldn't build an image based on `python:3.6` due to invalid ELF header for `libisl.so.10` during `pip install`. I reset everything again and got past that error.
1. During image builds, I saw a bunch of ELF errors fly by.
1. I started seeing segfaults on running tests (using pytest) in my test containers.

I ran the tests using the same images on a Linux host with no problems.

I reverted back to using `Docker.qcow2` and everything was working again.

* macOS 10.13.2
* Docker version 17.12.0-ce, build c97c6d6

**Comment by EugenMayer** [comment]
@vovimayhem interesting, so instead fixing this, they introduced a new one, now even breaking named/anon volumes which are only linux based mounts by tempering with the xhyve HV so far, that the scsi/sata controller implementation + raw is now even broken. 

Looks like a plan to me. The *.raw thing has far more issues then that. It breaks FS Event propagation too ( after a very short time )

**Comment by djs55** [comment]
I think there are some misunderstandings about the relationship between `Docker.raw`, `Docker.qcow2` and filesystem events.

Containers and images are stored on `overlay2` running inside the VM on top of a block device provided by the host. The `Docker.qcow2` or `Docker.raw` is used as the backing file for the block device. Filesystem events is entirely handled by the Linux kernel in this case.

Volumes shared from the host are mounted remotely using a protocol that is quite similar to NFS, extended with support for transporting filesystem events. The Mac application registers for filesystem events on the host, sends them to the VM where they are injected to Linux. The block device backed by `Docker.qcow2` or `Docker.raw` is not involved in this at all.

I realise this ticket is more of a discussion thread than a bug report :) but when you do hit a bug, please create a separate ticket unless you are sure that it is exactly the same as an existing one-- otherwise genuine problems can be lost in the discussion of other issues. Please also upload a diagnostic report and include repro instructions. Thanks a lot for your help!

Since this ticket isn't actually a bug report I'll close it. Feel free to keep chatting though and I'll try to keep an eye on the discussion.

**Comment by EugenMayer** [comment]
@djs55 i admit, this become rather a "frustration dump" ticket - i am part of it, so pointing at myself too. I think you are right to split out the actual issue tracking and the "real work" and step by step progress. I created https://github.com/docker/for-mac/issues/2417 for the issue i am referring to.

**Comment by craiga** [comment]
Just to be clear—does Docker perform well (enough) on High Sierra (under "normal" circumstances)?

**Comment by craiga** [comment]
OK, let me rephrase my question—does Docker perform well on High Sierra **compared with Sierra**?

**Comment by djs55** [comment]
@EugenMayer btw thanks for your work on docker-sync -- I'm hoping to be able to work more on volume sharing this quarter. I've already got patches which improve throughput, but I've not investigated the events yet. I'm currently focused on fixing some issues in the 17.12 on the networking side, but once those are done I'll switch back to file sharing.

**Comment by CvX** [comment]
@craiga that's only a single data point but yes - all **projects on my recently updated to High Sierra mac perform just as they did on Sierra**.
Most of them use native volume mounting, a few use docker-sync, just as before. No format changes, no container rebuilding.
